src/util/schnorr.rs

```diff
@@ -75,9 +79,71 @@ impl TweakedPublicKey {
     pub fn new(key: PublicKey) -> TweakedPublicKey {
         TweakedPublicKey(key)
     }
-    
-    /// Returns the underlying public key 
+
+    /// Returns the underlying public key
     pub fn into_inner(self) -> PublicKey {
         self.0
     }
+}
+
+/// A BIP340-341 serialized schnorr signature with the corresponding hash type.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct SchnorrSig {
+    /// The underlying schnorr signature
+    pub sig: secp256k1::schnorrsig::Signature,
+    /// The corresponding hash type
+    pub hash_ty: SchnorrSigHashType,
+}
+
+impl SchnorrSig {
+
+    /// Deserialize from slice
+    pub fn from_slice(sl: &[u8]) -> Result<Self, SchnorrSigError> {
+        match sl.split_last() {
+            Some((hash_ty, sig)) => {
+                let hash_ty = SchnorrSigHashType::from_u8(*hash_ty)
+                    .map_err(|_| SchnorrSigError::InvalidSighashType(*hash_ty))?;
+                let sig = secp256k1::schnorrsig::Signature::from_slice(sig)?;
+                Ok(SchnorrSig { sig, hash_ty })
+            },
+            None => {
+                Err(SchnorrSigError::Secp256k1(secp256k1::Error::InvalidSignature))?
+            },
+        }
+    }
+
+    /// Serialize SchnorrSig
+    pub fn serialize(&self) -> Vec<u8> {
```

Comment:
The same question with implementing `Encode`/`Decode` as above

Reply:
As mentioned in the above PR description, I did not want to clutter this PR with additional stuff. And there are a few subtle reasons why I did not implement this, or whether the Encodable even makes sense.

Encodable defines a bitcoin network serialization format. And signatures are always sent as witness data with serialization implemented for `Vec<Vec<u8>>`. This is the same reason why I did not define Encode/Decode trait methods for `ControlBlock`, `TaprootMerkleRoot` in #677 because they are a part of witness data. It might even create confusion whether the `Encodeable` includes the VarInt prefix in witness format or not. I was not involved in early rust-bitcoin days, but I presume this is the same reason why `Encode\Deocde` is not done for `bitcoin::PublicKey`

This is why I chose to implement methods directly on the type